### ISSUE TITLE
Lost oriention for jpg with options  SDWebImageScaleDownLargeImages

### ISSUE THREAD

**Issue by luobin23628** [rs/SDWebImage#2026]
### New Issue Checklist

* [x] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [x] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [x] I have searched for a similar issue in the [project](https://github.com/rs/SDWebImage/issues) and found none

### Issue Info

 Info                    | Value                               |
-------------------------|-------------------------------------|
 Platform Name           |  ios 
 Platform Version        |  8.0
 SDWebImage Version      | 4.0.0、4.1.0
 Integration Method      | cocoapods
 Xcode Version           | Xcode 7.3
 Repro rate              | all the time (100%)

### Issue Description and Steps

when i set  options SDWebImageScaleDownLargeImages, I lost exifs include oriention for jpg. It occur when convert jpg into png.

```
                    if (self.shouldDecompressImages) {
                        if (self.options & SDWebImageDownloaderScaleDownLargeImages) {
#if SD_UIKIT || SD_WATCH
                            image = [UIImage decodedAndScaledDownImageWithImage:image];
                            [self.imageData setData:UIImagePNGRepresentation(image)];
#endif
                        } else {
                            image = [UIImage decodedImageWithImage:image];
                        }
                    }
```

**Comment by dreampiggy** [comment]
`UIImagePNGRepresentation` and `UIImageJPEGRepresentation` will erase your EXIF information when convert `UIImage`(with non-up orientation) to `NSData`. It seems that an UIKit's bug but Apple adopt this behavior from the beginning.

Need some work to copy the EXIF metadata to that `NSData`

**Comment by dreampiggy** [comment]
Strange. That `decodedAndScaledDownImageWithImage` will not decoded alpha images and set non-alpha channel for CGBitmapContext. But then they use `UIImagePNGRepresentation` ? Maybe it's for better performance. What I need is to copy the EXIF information and copy to the data. But this may have some issue about different image format. I try to figure this.

**Comment by dreampiggy** [comment]
I provide a fix. Which drop the usage of that `UIImagePNGRepresentation` and encode UIImage to NSData with Core Graphics. It will copy all the EXIF, JFIF metadata and provide cross-platform(not just iOS/watchOS) compatibility.

The related PR is here: #2034

**Comment by dreampiggy** [comment]
Another way to fix. We now drop that behavior to modify this original image data 😅 . We just scale down the UIImage to reduce memory usage. The image data will not be cached  into memory and will be released after completion block. If you prefer to add some extra compress, you can grab the image data by yourself and do it.

**Comment by bpoplauschi** [comment]
@dreampiggy I'm just looking into this. So we fixed the issue of loosing the orientation. That is good.

But I think we need to also change the `imageData`. If we change `image` we should also change `imageData`, otherwise they are out of sync and can cause problems. Basically, I would consider we just got a smaller image than the one we actually did, so from that point forward, everything should be done using the smaller image.

**Comment by dreampiggy** [comment]
@bpoplauschi If we need to modify the original image data. We need to modify the `data` pointer as well. We can use that `encodedDataWithImage` since it now keep the EXIF information from `UIImage.imageOrientation`

**Comment by bpoplauschi** [comment]
@dreampiggy I'm not sure I follow. I am referring to
```
- (UIImage *)decompressedImageWithImage:(UIImage *)image
                                   data:(NSData **)data
                                options:(NSDictionary<NSString*, NSObject*>*)optionsDict
```
this already uses an in/out data param, but we just don't change that data inside the method implementation from `SDWebImageImageIOCoder`. That is my point. We should just change it inside, when we recompute the image. Indeed for getting NSData we can use the mechanism you described.

I don't think we need an option here because there is one already: `SDWebImageDownloaderScaleDownLargeImages`. If this is set explicitly by the user, then we will scale down the image, so it makes sense for the data returned to be in sync with that image.

**Comment by bpoplauschi** [comment]
@dreampiggy this is the last issue currently assigned to 4.2.0 release. Any updates here?

**Comment by dreampiggy** [comment]
@bpoplauschi Forgot this one. We need to modify that data pointer as well. I pick up a commit.

**Comment by bpoplauschi** [comment]
Initially fixed by #1991 + latest comments (about changing the `NSData` when scaling down images) fixed by #2074.
